File: LeetCode/Problems/1. Two Sum.py
#Old Implementation
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        if len(nums) == 0:
            return -1
        complement = {}
        for i, n in enumerate(nums):
            comp = target - n
            if comp not in complement:# and i != complement[target - n]:
                complement[n] = i
            else:
                return [complement[comp], i]



#One-pass hash table
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        #build a dict storing index and complements
        #iterate through the list and check if we've seen the complement
        if not nums:
            return []
        comp = {}
        
        for i,n in enumerate(nums):
            if n in comp:
                return [comp[n],i]
            else:
                comp[target-n] = i
                

File: LeetCode/Problems/C++/1. Two Sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> lookup;
        vector<int> ans;
        for(int i=0 ; i<nums.size();i++){
            if(lookup.find(target-nums[i]) == lookup.end()){
                lookup[nums[i]] = i;
            }
            else{
                ans.push_back(lookup[target-nums[i]]);
                ans.push_back(i);
                break;
            }
        }
        return ans;
    }
};


class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> diff;
        vector<int> v;
        for(int i = 0; i< nums.size();++i){ 
            
            if(diff.find(target - nums[i])== diff.end()){
                diff[nums[i]] = i;
            }
            else{
                v = {diff[target - nums[i]], i};
                break;
            }
        }
        return v;
    }
};